adsmp/tasks.py

```diff
@@ -74,57 +78,77 @@ def task_route_record(bibcode, force=False, delayed=1):
-    if processed is None:
-        # It was never sent to Solr
-        processed = adsputils.get_date(year_zero)
-
-    is_complete = all([bib_data_updated, orcid_claims_updated, nonbib_data_updated])
-
-    if is_complete:
-        if force is False and all([bib_data_updated and bib_data_updated < processed,
-               orcid_claims_updated and orcid_claims_updated < processed,
-               nonbib_data_updated and nonbib_data_updated < processed]):
-            logger.debug('Nothing to do, it was already indexed/processed')
-            return
-        else:
-            # build the record and send it to solr
-            logger.debug('Updating solr')
-            solr_doc = solr_updater.transform_json_record(r)
-            solr_updater.update_solr(solr_doc, app.conf.get('SOLR_URLS'))
-            app.update_processed_timestamp(bibcode)
-    else:
-        # if we have at least the bib data, index it
-        if force is True and bib_data_updated:
-            logger.warn('Forced indexing of: %s (metadata=%s, orcid=%s, nonbib=%s, fulltext=%s)' % \
-                        (bibcode, bib_data_updated, orcid_claims_updated, nonbib_data_updated, fulltext_updated))
-            # build the record and send it to solr
-            solr_doc = solr_updater.transform_json_record(r)
-            solr_updater.update_solr(solr_doc, app.conf.get('SOLR_URLS'))
-            app.update_processed_timestamp(bibcode)
+    for bibcode in bibcodes:
+        r = app.get_record(bibcode)
+    
+        if r is None:
+            logger.error('The bibcode %s doesn\'t exist!', bibcode)
+            continue
+    
+        bib_data_updated = r.get('bib_data_updated', None)
+        orcid_claims_updated = r.get('orcid_claims_updated', None)
+        nonbib_data_updated = r.get('nonbib_data_updated', None)
+        fulltext_updated = r.get('fulltext_updated', None)
+        metrics_updated = r.get('metrics_updated', None)
+    
+        year_zero = '1972'
+        processed = r.get('processed', adsputils.get_date(year_zero))
+        if processed is None:
+            # It was never sent to Solr
+            processed = adsputils.get_date(year_zero)
+    
+        is_complete = all([bib_data_updated, orcid_claims_updated, nonbib_data_updated])
+    
+        if is_complete:
+            if force is False and all([bib_data_updated and bib_data_updated < processed,
+                   orcid_claims_updated and orcid_claims_updated < processed,
+                   nonbib_data_updated and nonbib_data_updated < processed]):
+                logger.debug('Nothing to do for %s, it was already indexed/processed', bibcode)
+                continue
+            else:
+                # build the solr record
+                batch.append(solr_updater.transform_json_record(r))
+                # get data for metrics
+                m = r.get('metrics', None)
+                if m:
+                    if r.get('processed'):
```

Comment:
I assume checking processed tells us if the record was written before.  That isn't definitive on whether we should do a sql insert or update.  Perhaps we processed the record and then database died.  Perhaps after updating a metrics record we had a database error that prevented processed from being set.  Can we rely on our processed flag and the metrics database staying in sync?